src/common/vector_operations/vector_cast.cpp

```diff
@@ -168,10 +168,126 @@ static bool NumericCastSwitch(Vector &source, Vector &result, idx_t count, strin
+	bool all_converted = true;
+	if (source_mask.AllValid()) {
+		for (idx_t i = 0; i < count; i++) {
+			auto string_value = source_data[i].GetString();
+			auto pos = EnumType::GetPos(result_type, string_value);
+			if (pos == -1) {
+				if (!error_message) {
+					throw CastException(source_type, result_type);
+				} else {
+					all_converted = false;
+					result_mask.SetInvalid(i);
+				}
+			} else {
+				result_data[i] = pos;
+			}
+		}
+	} else {
+		for (idx_t i = 0; i < count; i++) {
+			if (source_mask.RowIsValid(i)) {
+				auto string_value = source_data[i].GetString();
+				auto pos = EnumType::GetPos(result_type, string_value);
+				if (pos == -1) {
+					if (!error_message) {
+						throw CastException(source_type, result_type);
+					} else {
+						all_converted = false;
+						result_mask.SetInvalid(i);
+					}
+				} else {
+					result_data[i] = pos;
+				}
+			} else {
+				result_mask.SetInvalid(i);
+			}
+		}
+	}
+	return all_converted;
+}
+
+template <class T>
+bool TransformEnum(Vector &source, Vector &result, idx_t count, string *error_message) {
+	D_ASSERT(source.GetType().id() == LogicalTypeId::VARCHAR);
+	auto enum_name = EnumType::GetTypeName(result.GetType());
+	bool all_converted = true;
+	switch (source.GetVectorType()) {
+	case VectorType::CONSTANT_VECTOR: {
+		result.SetVectorType(VectorType::CONSTANT_VECTOR);
+		auto source_data = ConstantVector::GetData<string_t>(source);
+		auto result_data = ConstantVector::GetData<T>(result);
+		if (ConstantVector::IsNull(source)) {
+			ConstantVector::SetNull(result, true);
+		} else {
+			ConstantVector::SetNull(result, false);
+			auto string_value = source_data[0].GetString();
+			auto pos = EnumType::GetPos(result.GetType(), source_data[0].GetString());
+			if (pos == -1) {
+				if (!error_message) {
+					throw CastException(source.GetType(), result.GetType());
+				} else {
+					all_converted = false;
+					ConstantVector::SetNull(result, true);
+				}
+			} else {
+				result_data[0] = pos;
+			}
+		}
+		break;
+	}
+	case VectorType::FLAT_VECTOR: {
+		result.SetVectorType(VectorType::FLAT_VECTOR);
+
+		auto source_data = FlatVector::GetData<string_t>(source);
+		auto source_mask = FlatVector::Validity(source);
+
+		auto result_data = FlatVector::GetData<T>(result);
+		auto result_mask = FlatVector::Validity(result);
+
+		all_converted = FillEnum(source_data, source_mask, source.GetType(), result_data, result_mask, result.GetType(),
+		                         count, error_message);
+		break;
+	}
+	default: {
+		VectorData vdata;
+		source.Orrify(count, vdata);
+
+		result.SetVectorType(VectorType::FLAT_VECTOR);
+
+		auto source_data = (string_t *)vdata.data;
+		auto source_mask = vdata.validity;
+
+		auto result_data = FlatVector::GetData<T>(result);
+		auto result_mask = FlatVector::Validity(result);
+
+		all_converted = FillEnum(source_data, source_mask, source.GetType(), result_data, result_mask, result.GetType(),
```

Comment:
This seems incorrect, the selection vector is not used. Can we add extra tests with casts to enum with dictionary vectors (e.g. conditions in a WHERE clause that remove tuples)?